lib/twocohom.gi

```diff
@@ -1518,17 +1519,125 @@ local r,z,ogens,n,gens,str,dim,i,j,f,rels,new,quot,g,p,collect,m,e,fp,old,sim,
       fi;
     od;
     return new;
-
   end;
 
-
+  melmvec:=function(v)
+  local i,a,w;
+    w:=One(f);
+    for i in [1..Length(v)] do
+      a:=Int(v[i]);
+      if prime=2 or a*2<prime then
+        w:=w*gens[2*i-1+mn]^a;
+      else
+        a:=prime-a;
+        w:=w*gens[2*i+mn]^a;
+      fi;
+    od;
+    return w;
+  end;
 
   r:=arg[1];
   z:=arg[2];
   ogens:=GeneratorsOfGroup(r.presentation.group);
+  dim:=r.module.dimension;
+  prime:=Size(r.module.field);
+
+  if ValueOption("normalform")<>true then
+    mon:=fail;
+  else
+    # Construct the monoid and rewriting system, as this is cheap to do but
+    # will allow for reduced multiplication. Do so before the group, so there is no
+    # issue about overwriting variables that might be needed later
+    mon:=Image(r.monhom);
+    str:=List(GeneratorsOfMonoid(mon),String);
+    mn:=Length(str);
+    for i in [1..dim] do
+      Add(str,Concatenation("m",String(i)));
+      Add(str,Concatenation("m",String(i),"^-1"));
+    od;
+    f:=FreeMonoid(str);
+    gens:=GeneratorsOfMonoid(f);
```

Comment:
A lot of this code is not exercised by any tests (as indicated by the Codecov markers on https://github.com/gap-system/gap/pull/4441/files). Is there any chance to add at least some "cheap" one?

Reply:
Done in manual example